File: 10-01-2025_Count_distinct_elements_in_every_window.py
from collections import defaultdict

class Solution:
    def countDistinct(self, arr, k):
        ans = []
        freq = defaultdict(int)  # Dictionary to store frequencies of elements in the current window
        distinct_count = 0

        # Initialize the first window
        for i in range(k):
            if freq[arr[i]] == 0:
                distinct_count += 1
            freq[arr[i]] += 1
        ans.append(distinct_count)

        # Slide the window
        for i in range(k, len(arr)):
            # Remove the element going out of the window
            if freq[arr[i - k]] == 1:
                distinct_count -= 1
            freq[arr[i - k]] -= 1

            # Add the new element coming into the window
            if freq[arr[i]] == 0:
                distinct_count += 1
            freq[arr[i]] += 1

            ans.append(distinct_count)

        return ans

File: 10-01-2025_Count_distinct_elements_in_every_window.java
import java.util.*;

class Solution {
    public ArrayList<Integer> countDistinct(int[] arr, int k) {
        ArrayList<Integer> ans = new ArrayList<>();
        HashMap<Integer, Integer> hash = new HashMap<>();
        int n = arr.length;
        int distinctCount = 0;

        // Initialize the first window
        for (int i = 0; i < k; i++) {
            if (!hash.containsKey(arr[i])) {
                distinctCount++;
            }
            hash.put(arr[i], hash.getOrDefault(arr[i], 0) + 1);
        }
        ans.add(distinctCount);

        // Slide the window
        for (int i = k; i < n; i++) {
            // Remove the element going out of the window
            if (hash.get(arr[i - k]) == 1) {
                distinctCount--;
            }
            hash.put(arr[i - k], hash.get(arr[i - k]) - 1);

            // Add the new element coming into the window
            if (!hash.containsKey(arr[i]) || hash.get(arr[i]) == 0) {
                distinctCount++;
            }
            hash.put(arr[i], hash.getOrDefault(arr[i], 0) + 1);

            ans.add(distinctCount);
        }

        return ans;
    }
}
